lib/ansible/utils/ssh_agent.py

```diff
@@ -0,0 +1,660 @@
+        ...
+
+
+class VariableSized:
+    @classmethod
+    def from_blob(cls, blob: memoryview | bytes) -> t.Self:
+        raise NotImplementedError
+
+    @classmethod
+    def consume_from_blob(cls, blob: memoryview | bytes) -> tuple[t.Self, memoryview | bytes]:
+        length = uint32.from_blob(blob[:4])
+        blob = blob[4:]
+        return cls.from_blob(blob[:length]), blob[length:]
+
+
+class uint32(int):
+    def to_blob(self) -> bytes:
+        return self.to_bytes(length=4, byteorder='big')
+
+    @classmethod
+    def from_blob(cls, blob: memoryview | bytes) -> t.Self:
+        return cls.from_bytes(blob, byteorder='big')
+
+    @classmethod
+    def consume_from_blob(cls, blob: memoryview | bytes) -> tuple[t.Self, memoryview | bytes]:
+        length = uint32(4)
+        return cls.from_blob(blob[:length]), blob[length:]
+
+
+class mpint(int, VariableSized):
+    def to_blob(self) -> bytes:
+        if self < 0:
+            raise ValueError("negative mpint not allowed")
+        if not self:
+            return b""
+        nbytes = (self.bit_length() + 8) // 8
+        ret = bytearray(self.to_bytes(length=nbytes, byteorder='big'))
+        ret[:0] = uint32(len(ret)).to_blob()
+        return ret
+
+    @classmethod
+    def from_blob(cls, blob: memoryview | bytes) -> t.Self:
+        if blob and blob[0] > 127:
+            raise ValueError("Invalid data")
+        return cls.from_bytes(blob, byteorder='big')
+
+
+class constraints(bytes):
+    def to_blob(self) -> bytes:
+        return self
+
+
+class binary_string(bytes, VariableSized):
+    def to_blob(self) -> bytes:
+        if length := len(self):
+            return uint32(length).to_blob() + self
+        else:
+            return b""
+
+    @classmethod
+    def from_blob(cls, blob: memoryview | bytes) -> t.Self:
+        return cls(blob)
+
+
+class unicode_string(str, VariableSized):
+    def to_blob(self) -> bytes:
+        val = self.encode('utf-8')
+        if length := len(val):
+            return uint32(length).to_blob() + val
+        else:
+            return b""
+
+    @classmethod
+    def from_blob(cls, blob: memoryview | bytes) -> t.Self:
+        return cls(bytes(blob).decode('utf-8'))
+
+
+class KeyAlgo(str, VariableSized, enum.Enum):
+    RSA = "ssh-rsa"
+    DSA = "ssh-dss"
+    ECDSA256 = "ecdsa-sha2-nistp256"
+    SKECDSA256 = "sk-ecdsa-sha2-nistp256@openssh.com"
+    ECDSA384 = "ecdsa-sha2-nistp384"
+    ECDSA521 = "ecdsa-sha2-nistp521"
+    ED25519 = "ssh-ed25519"
+    SKED25519 = "sk-ssh-ed25519@openssh.com"
+    RSASHA256 = "rsa-sha2-256"
+    RSASHA512 = "rsa-sha2-512"
+
+    @property
+    def main_type(self):
+        match self:
+            case self.RSA:
+                return 'RSA'
+            case self.DSA:
+                return 'DSA'
+            case self.ECDSA256 | self.ECDSA384 | self.ECDSA521:
+                return 'ECDSA'
+            case self.ED25519:
+                return 'ED25519'
+            case _:
```

Comment:
Do we expect that this never happens?
```suggestion
            case _ as unreachable:
                t.assert_never(unreachable)
```

Reply:
I feel like raising `NotImplementedError` sends a better message to the user as to what is going on 🤷🏻‍♂️

Reply:
@mkrizek it's not for the user, I'm not suggesting dropping it. But for typing, it's useful to mark the branch with `assert_never()`.

Reply:
> it's not for the user, I'm not suggesting dropping it

The change you are suggesting renders raising `NotImplementedError` unreachable, no? `t.assert_never(unreachable)` would raise `AssertionError` at runtime.

> But for typing, it's useful to mark the branch with assert_never().

Would that be actually useful in this case? The parsing is all dynamic performed on responses from `ssh-agent`, so the static checker won't be able to detect this anyway?

Reply:
> > it's not for the user, I'm not suggesting dropping it
> 
> The change you are suggesting renders raising `NotImplementedError` unreachable, no? `t.assert_never(unreachable)` would raise `AssertionError` at runtime.

Are you sure? I thought it's no-op.

> 
> > But for typing, it's useful to mark the branch with assert_never().
> 
> Would that be actually useful in this case? The parsing is all dynamic performed on responses from `ssh-agent`, so the static checker won't be able to detect this anyway?

Still useful, but you'd need to make this more exhaustive: https://typing.readthedocs.io/en/latest/guides/unreachable.html

Reply:
Ah, I didn't realize that it's an `AssertionError` at runtime. So it'd replace raising notimplemented...

Reply:
> I feel like raising `NotImplementedError` sends a better message to the user as to what is going on 🤷🏻‍♂️

So I wanted to add that type checking makes it so it wouldn't even reach the user in the first place. Unlike raising validation errors, this is a mechanism for making sure that the developers never call it like this, not the users.

Are there any tests that demonstrate behavior in this code branch?